src/libsd.s

```diff
@@ -1,299 +1,367 @@
+  ldy #4
+  lda (zp_sd_address),y    ; data 4
+  jsr sd_writebyte
+  ldy #5
+  lda (zp_sd_address),y    ; crc
+  jsr sd_writebyte
+
+  jsr sd_waitresult
+  pha
+
+  ; Debug print the result code
+  jsr print_hex
+
+  ; End command
+  lda #SD_CS | SD_MOSI   ; set CS high again
+  sta PORTA
+
+  pla   ; restore result code
+  rts
+
+
+sd_readsector:
+  ; Read a sector from the SD card.  A sector is 512 bytes.
+  ;
+  ; Parameters:
+  ;    zp_sd_currentsector   32-bit sector number
+  ;    zp_sd_address     address of buffer to receive data
+  
+  lda #SD_MOSI
+  sta PORTA
+
+  ; Command 17, arg is sector number, crc not checked
+  lda #$51                    ; CMD17 - READ_SINGLE_BLOCK
+  jsr sd_writebyte
+  lda zp_sd_currentsector+3   ; sector 24:31
+  jsr sd_writebyte
+  lda zp_sd_currentsector+2   ; sector 16:23
+  jsr sd_writebyte
+  lda zp_sd_currentsector+1   ; sector 8:15
+  jsr sd_writebyte
+  lda zp_sd_currentsector     ; sector 0:7
+  jsr sd_writebyte
+  lda #$01                    ; crc (not checked)
+  jsr sd_writebyte
+
+  jsr sd_waitresult
+  cmp #$00
+  bne sd_fail
+
+  ; wait for data
+  jsr sd_waitresult
+  cmp #$fe
+  bne sd_fail
+
+  ; Need to read 512 bytes - two pages of 256 bytes each
+  jsr .readpage
+  inc zp_sd_address+1
+  jsr .readpage
+  dec zp_sd_address+1
+
+  ; End command
+  lda #SD_CS | SD_MOSI
+  sta PORTA
+
+  rts
+
+.readpage
+  ; Read 256 bytes to the address at zp_sd_address
+  ldy #0
+.readloop
+  jsr sd_readbyte
+  sta (zp_sd_address),y
+  iny
+  bne .readloop
+  rts
+
+sd_fail:
+  lda #'s'
+  jsr print_char
+  lda #':'
+  jsr print_char
+  lda #'f'
+  jsr print_char
+.failloop
+  jmp .failloop
+
+
+sd_writesector:
+  ; Write a sector to the SD card.  A sector is 512 bytes.
+  ;
+  ; Parameters:
+  ;    zp_sd_currentsector   32-bit sector number
+  ;    zp_sd_address     address of buffer to take data from
+  
+  lda #SD_MOSI
+  sta PORTA
+
+  ; Command 24, arg is sector number, crc not checked
+  lda #$58                    ; CMD24 - WRITE_BLOCK
+  jsr sd_writebyte
+  lda zp_sd_currentsector+3   ; sector 24:31
+  jsr sd_writebyte
+  lda zp_sd_currentsector+2   ; sector 16:23
+  jsr sd_writebyte
+  lda zp_sd_currentsector+1   ; sector 8:15
+  jsr sd_writebyte
+  lda zp_sd_currentsector     ; sector 0:7
+  jsr sd_writebyte
+  lda #$01                    ; crc (not checked)
+  jsr sd_writebyte
+
+  jsr sd_waitresult
+  cmp #$00
+  bne sd_fail
+
+  ; Send start token
+  lda #$fe
+  jsr sd_writebyte
+
+  ; Need to write 512 bytes - two pages of 256 bytes each
+  jsr .writepage
+  inc zp_sd_address+1
+  jsr .writepage
+  dec zp_sd_address+1
+
+  ; wait for data response
+  jsr sd_waitresult
+  and #$1f
+  cmp #$05
+  bne sd_fail
+
+.waitidle
+  jsr sd_readbyte
+  cmp #$ff
+  bne .waitidle
+
+  ; End command
+  lda #SD_CS | SD_MOSI ; set cs and mosi high (disconnected)
+  sta PORTA
+
+  rts
+
+.writepage
+  ; Write 256 bytes fom zp_sd_address
+  ldy #0
+.writeloop
+  lda (zp_sd_address),y
+  tya

```

Comment:
did you notice this bug?  A is loaded to write and then is immediately stomped on by Y.  As such, all writes are just writing 00-FF for the whole sector.   Probably should fix this one too!

Reply:
another one that was fixed in the upstream xpl-32 but not here...